File: src/lua/bindings/path.cpp
#pragma once
#include "path.hpp"

namespace lua::path
{
	// Lua API: Table
	// Name: path
	// Table containing helpers for manipulating file or directory paths

	// Lua API: Function
	// Table: path
	// Name: combine
	// Param: path: string: Any amount of string is accepted.
	// Returns: string: Returns the combined path
	// Combines strings into a path.
	static std::string combine(sol::variadic_args args)
	{
		try
		{
			std::filesystem::path res{};

			for (const auto& arg : args)
			{
				if (arg.is<std::string>())
				{
					if (res.empty())
					{
						res = arg.as<std::string>();
					}
					else
					{
						res /= arg.as<std::string>();
					}
				}
			}

			return res.string();
		}
		catch (const std::exception& e)
		{
			LOG(WARNING) << e.what();
		}

		return "";
	}

	// Lua API: Function
	// Table: path
	// Name: get_parent
	// Param: path: string: The path for which to retrieve the parent directory.
	// Returns: string: Returns the parent path
	// Retrieves the parent directory of the specified path, including both absolute and relative paths.
	static std::string get_parent(const std::string& path)
	{
		try
		{
			std::filesystem::path res = path;
			return (char*)res.parent_path().u8string().c_str();
		}
		catch (const std::exception& e)
		{
			LOG(WARNING) << e.what();
		}

		return "";
	}

	// Lua API: Function
	// Table: path
	// Name: get_directories
	// Param: root_path: string: The path to the directory to search.
	// Returns: string table: Returns the names of subdirectories under the given root_path
	static std::vector<std::string> get_directories(const std::string& root_path)
	{
		std::vector<std::string> res;

		try
		{
			for (const auto& entry : std::filesystem::recursive_directory_iterator(root_path, std::filesystem::directory_options::skip_permission_denied))
			{
				if (!entry.is_directory())
				{
					continue;
				}

				res.push_back((char*)entry.path().u8string().c_str());
			}
		}
		catch (const std::exception& e)
		{
			LOG(WARNING) << e.what();
		}

		return res;
	}

	// Lua API: Function
	// Table: path
	// Name: get_files
	// Param: root_path: string: The path to the directory to search.
	// Returns: string table: Returns the names of all the files under the given root_path
	static std::vector<std::string> get_files(const std::string& root_path)
	{
		std::vector<std::string> res;

		try
		{
			for (const auto& entry : std::filesystem::recursive_directory_iterator(root_path, std::filesystem::directory_options::skip_permission_denied))
			{
				if (entry.is_directory())
				{
					continue;
				}

				res.push_back((char*)entry.path().u8string().c_str());
			}
		}
		catch (const std::exception& e)
		{
			LOG(WARNING) << e.what();
		}

		return res;
	}

	// Lua API: Function
	// Table: path
	// Name: filename
	// Param: path: string: The path for which to retrieve the filename.
	// Returns: string: Returns the filename identified by the path.
	static std::string filename(const std::string& path)
	{
		try
		{
			std::filesystem::path res = path;
			return (char*)res.filename().u8string().c_str();
		}
		catch (const std::exception& e)
		{
			LOG(WARNING) << e.what();
		}

		return "";
	}

	// Lua API: Function
	// Table: path
	// Name: stem
	// Param: path: string: The path for which to retrieve the stem.
	// Returns: string: Returns the stem of the filename identified by the path (i.e. the filename without the final extension).
	static std::string stem(const std::string& path)
	{
		try
		{
			std::filesystem::path res = path;
			return (char*)res.stem().u8string().c_str();
		}
		catch (const std::exception& e)
		{
			LOG(WARNING) << e.what();
		}

		return "";
	}

	void bind(sol::state& state)
	{
		auto ns               = state["path"].get_or_create<sol::table>();
		ns["combine"]         = combine;
		ns["get_parent"]      = get_parent;
		ns["get_directories"] = get_directories;
		ns["get_files"]       = get_files;
		ns["filename"]        = filename;
		ns["stem"]            = stem;
	}
} // namespace lua::path

File: docs/lua/tables/definitions/path.lua
---@meta path

-- Table containing helpers for manipulating file or directory paths
---@class (exact) path

-- Combines strings into a path.
---@param path string Any amount of string is accepted.
---@return string # Returns the combined path
function path.combine(path) end

-- Retrieves the parent directory of the specified path, including both absolute and relative paths.
---@param path string The path for which to retrieve the parent directory.
---@return string # Returns the parent path
function path.get_parent(path) end

---@param root_path string The path to the directory to search.
---@return string table # Returns the names of subdirectories under the given root_path
function path.get_directories(root_path) end

---@param root_path string The path to the directory to search.
---@return string table # Returns the names of all the files under the given root_path
function path.get_files(root_path) end

---@param path string The path for which to retrieve the filename.
---@return string # Returns the filename identified by the path.
function path.filename(path) end

---@param path string The path for which to retrieve the stem.
---@return string # Returns the stem of the filename identified by the path (i.e. the filename without the final extension).
function path.stem(path) end

---@param path string The path to the new directory to create.
---@return boolean # true if a directory was newly created for the directory p resolves to, false otherwise.
function path.create_directory(path) end

---@param path string The path to check.
---@return boolean # true if the path exists, false otherwise.
function path.exists(path) end

-- Registers a callback that will be called when a file changes.
--**Example Usage:**
--```lua
--path.add_file_watcher(_ENV["!config_mod_folder_path"], function (file_name, timestamp)
--		log.info(file_name, timestamp)
--end)
--```
---@param path string The path to add file watcher.
---@param callback function callback that match signature function ( file_name, timestamp ).
function path.add_file_watcher(path, callback) end


